src/axom/primal/geometry/Hexahedron.hpp

```diff
@@ -0,0 +1,376 @@
+   */
+  AXOM_HOST_DEVICE PointType& operator[](int idx)
+  {
+    SLIC_ASSERT(idx >= 0 && idx < NUM_HEX_VERTS);
+    return m_points[idx];
+  }
+
+  /*!
+   * \brief Index operator to get the i^th vertex
+   * \param idx The index of the desired vertex
+   * \pre idx is 0, 1, 2, 3, 4, 5, 6, or 7
+   */
+  AXOM_HOST_DEVICE const PointType& operator[](int idx) const
+  {
+    SLIC_ASSERT(idx >= 0 && idx < NUM_HEX_VERTS);
+    return m_points[idx];
+  }
+
+  /*!
+   * \brief Computes the average of the hexahedron's vertex
+   *  positions
+   *
+   * \return A point at the mean of the hexahedron's vertices
+   */
+  AXOM_HOST_DEVICE
+  PointType vertexMean() const
+  {
+    PointType sum;
+
+    for(int i = 0; i < NUM_HEX_VERTS; ++i)
+    {
+      sum.array() += m_points[i].array();
+    }
+    sum.array() /= NUM_HEX_VERTS;
+
+    return sum;
+  }
+
+  /**
+   * \brief Finds the signed volume of the hexahedron
+   *
+   * Uses algorithm in
+   * Grandy, J. Efficient computation of volume of hexahedral cells.
+   * United States: N. p., 1997. Web. doi:10.2172/632793.
+   *
+   * \return The volume of the hexahedron
+   *
+   * \note Indices are different from the original paper
+   */
+  AXOM_HOST_DEVICE
+  double signedVolume() const
+  {
+    constexpr double scale = 1. / 12.;
+    return scale *
+      (VectorType::scalar_triple_product(VectorType(m_points[6] - m_points[3]) +
+                                           VectorType(m_points[5] - m_points[0]),
+                                         VectorType(m_points[6] - m_points[4]),
+                                         VectorType(m_points[7] - m_points[0])) +
+       VectorType::scalar_triple_product(VectorType(m_points[5] - m_points[0]),
+                                         VectorType(m_points[6] - m_points[4]) +
+                                           VectorType(m_points[2] - m_points[0]),
+                                         VectorType(m_points[6] - m_points[1])) +
+       VectorType::scalar_triple_product(VectorType(m_points[6] - m_points[3]),
+                                         VectorType(m_points[2] - m_points[0]),
+                                         VectorType(m_points[6] - m_points[1]) +
+                                           VectorType(m_points[7] - m_points[0])));
+  }
+
+  /*!
+   * \brief Returns the absolute (unsigned) volume of the hexahedron
+   * \sa signedVolume()
+   */
+  AXOM_HOST_DEVICE
+  double volume() const { return axom::utilities::abs(signedVolume()); }
+
+  /**
+   * \brief Method to decompose a Hexahedron
+   *        into 24 Tetrahedrons.
+   *        Each Tetrahedron consists of 4 points:
+   *          (1) The mean of all Hexahedron points
+   *          (2,3) Two adjacent vertices on a Hexahedron face
+   *          (4) The mean of the current Hexahedron face
+   *
+   * \param tets [out] The tetrahedrons
+   *
+   * \note Assumes tets is pre-allocated
+   */
+  AXOM_HOST_DEVICE
+  void triangulate(TetrahedronType* tets)
```

Comment:
If the output each time is a fixed-size array, my gut feeling is that we should probably make that explicit in the API, either by:
- accepting a fixed reference to an array (`void triangulate(TetrahedronType (&tets)[24])`)
- or returning something like a `StackArray<TetrahedronType, 24>`